src/integration-test/java/com/commercetools/sync/integration/services/impl/CartDiscountServiceImplIT.java

```diff
@@ -0,0 +1,432 @@
+        final Set<String> cartDiscountKeys = new HashSet<>();
+        cartDiscountKeys.add("cart_discount_key_1");
+        cartDiscountKeys.add("cart_discount_key_2");
+
+        final Set<CartDiscount> matchingCartDiscounts =
+                cartDiscountService.fetchMatchingCartDiscountsByKeys(cartDiscountKeys)
+                        .toCompletableFuture()
+                        .join();
+
+        assertThat(matchingCartDiscounts).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchMatchingCartDiscountsByKeys_WithAnyExistingKeys_ShouldReturnASetOfCartDiscounts() {
+        final Set<String> cartDiscountKeys = new HashSet<>();
+        cartDiscountKeys.add(CART_DISCOUNT_KEY_1);
+
+        final Set<CartDiscount> matchingCartDiscounts =
+                cartDiscountService.fetchMatchingCartDiscountsByKeys(cartDiscountKeys)
+                        .toCompletableFuture()
+                        .join();
+
+        assertThat(matchingCartDiscounts).hasSize(1);
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchMatchingCartDiscountsByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
+        // Mock sphere client to return BadGatewayException on any request.
+        final SphereClient spyClient = spy(CTP_TARGET_CLIENT);
+        when(spyClient.execute(any(CartDiscountQuery.class)))
+                .thenReturn(exceptionallyCompletedFuture(new BadGatewayException()))
+                .thenCallRealMethod();
+
+        final CartDiscountSyncOptions spyOptions =
+                CartDiscountSyncOptionsBuilder.of(spyClient)
+                        .errorCallback((errorMessage, exception) -> {
+                            errorCallBackMessages.add(errorMessage);
+                            errorCallBackExceptions.add(exception);
+                        })
+                        .build();
+
+        final CartDiscountService spyCartDiscountService = new CartDiscountServiceImpl(spyOptions);
+
+
+        final Set<String> cartDiscountKeys = new HashSet<>();
+        cartDiscountKeys.add(CART_DISCOUNT_KEY_1);
+
+        // test and assert
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(spyCartDiscountService.fetchMatchingCartDiscountsByKeys(cartDiscountKeys))
+                .hasFailedWithThrowableThat()
+                .isExactlyInstanceOf(BadGatewayException.class);
+    }
+
+    @Test
+    void fetchCachedCartDiscountId_WithNotCachedExistingCartDiscount_ShouldFetchCartDiscount() {
+        final Optional<String> cartDiscountId = cartDiscountService.fetchCachedCartDiscountId(CART_DISCOUNT_KEY_1)
+                .toCompletableFuture()
+                .join();
+
+        assertThat(cartDiscountId).isNotEmpty();
```

Comment:
I have a feeling we are too relaxed on the asserting here. Don't you think we need to assert that this key is actually cached?

Reply:
by either checking the cache or by verifying that on a second fetch, no request is made.

Reply:
5cdd3f9